prow/tide/gerrit.go

```diff
@@ -0,0 +1,312 @@
+}
+
+// Query returns all PRs from configured gerrit org/repos.
+func (p *GerritProvider) Query() (map[string]CodeReviewCommon, error) {
+	// lastUpdate is used by gerrit adapter for achieving incremental query. In
+	// tide case we want to get everything so use default time.Time, which
+	// should be 1970,1,1.
+	var lastUpdate time.Time
+
+	res := make(map[string]CodeReviewCommon)
+	// This is querying serially, which would safely guard against quota issues.
+	// TODO(chaodai): parallize this to boot the performance.
+	for instance, projs := range p.cfg().Tide.Gerrit.Queries.AllRepos() {
+		for projName := range projs {
+			changes, err := p.gc.QueryChangesForProject(instance, projName, lastUpdate, p.cfg().Gerrit.RateLimit, []string{gerritDefaultQueryParam})
+			if err != nil {
+				p.logger.WithFields(logrus.Fields{"instance": instance, "project": projName}).WithError(err).Error("Querying gerrit project for changes.")
+				continue
+			}
+			for _, pr := range changes {
+				crc := CodeReviewCommonFromGerrit(&pr, instance)
+				res[prKey(crc)] = *crc
+			}
+		}
+	}
+
+	return res, nil
+}
+
+func (p *GerritProvider) blockers() (blockers.Blockers, error) {
+	// This is not supported yet, so return an empty blocekr for now.
+	return blockers.Blockers{}, nil
+}
+
+func (p *GerritProvider) isAllowedToMerge(crc *CodeReviewCommon) (string, error) {
+	if crc.Mergeable == string(githubql.MergeableStateConflicting) {
+		return "PR has a merge conflict.", nil
+	}
+	return "", nil
+}
+
+// GetRef gets the latest revision from org/repo/branch.
+func (p *GerritProvider) GetRef(org, repo, ref string) (string, error) {
+	return p.gc.GetBranchRevision(org, repo, ref)
+}
+
+// headContexts gets the status contexts for the commit with OID ==
+// pr.HeadRefOID
+//
+// Assuming all submission requirements are already met as the PRs queried are
+// already submittable. So the focus here is to ensure that all prowjobs were
+// tested against latest baseSHA.
+// Prow parses baseSHA from the `Description` field of a context, will make sure
+// that all Prow jobs that vote to required labels are represented here.
+func (p *GerritProvider) headContexts(crc *CodeReviewCommon) ([]Context, error) {
+	var res []Context
+
+	selector := map[string]string{
+		kube.GerritRevision:   crc.HeadRefOID,
+		kube.ProwJobTypeLabel: string(prowapi.PresubmitJob),
+		kube.OrgLabel:         crc.Org,
+		kube.RepoLabel:        crc.Repo,
+		kube.PullLabel:        strconv.Itoa(crc.Number),
+	}
+	var pjs v1.ProwJobList
+	if err := p.pjclientset.List(context.Background(), &pjs, ctrlruntimeclient.MatchingLabels(selector)); err != nil {
+		return nil, fmt.Errorf("Cannot list prowjob with selector %v", selector)
+	}
+
+	// keep track of latest prowjobs only
+	latestPjs := make(map[string]*prowapi.ProwJob)
+	for _, pj := range pjs.Items {
+		pj := pj
+		exist, ok := latestPjs[pj.Spec.Context]
+		if ok && exist.CreationTimestamp.After(pj.CreationTimestamp.Time) {
```

Comment:
Nit: convert to `if exist, ok := latestPjs[pj.Spec.Context]; ok && exist.... { ... }` style.

Reply:
done